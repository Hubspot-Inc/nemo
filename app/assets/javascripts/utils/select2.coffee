select2 = {
  getAjaxParams: (url, resultsKey = 'results', textKey = 'text') ->
    url: url
    dataType: 'json'
    delay: 250,
    data: (params) ->
      search: params.term
      page: params.page
    processResults: (data) ->
      results = data[resultsKey]
      results.forEach((r) -> r['text'] = r[textKey]) unless textKey == 'text'
      {results: results, pagination: {more: data.more}}
    cache: true
}

ELMO.select2 = select2
