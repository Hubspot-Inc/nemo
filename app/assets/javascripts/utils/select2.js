export function select2AjaxParams(url, resultsKey = 'results') {
  return {
    url,
    dataType: 'json',
    delay: 250,
    data: ({ term: search, page }) => ({
      search,
      page,
    }),
    processResults: ({ [resultsKey]: results, more }) => ({
      results,
      pagination: { more },
    }),
    cache: true,
  };
}
