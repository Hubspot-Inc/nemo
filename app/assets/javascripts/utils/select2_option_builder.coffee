# Builds options for select2 controls so that common options can be reused without duplication.
# It's designed like this, instead of as a wrapper of select2, so it can be used in React as well
# as Backbone.
class ELMO.Utils.Select2OptionBuilder
  ajax: (url, resultsKey = 'results', textKey = 'text') ->
    url: url
    dataType: 'json'
    delay: 250,
    data: (params) ->
      search: params.term
      page: params.page
    processResults: (data) ->
      results = data[resultsKey]
      results.forEach((r) -> r['text'] = r[textKey]) unless textKey == 'text'
      {results: results, pagination: {more: data.more}}
    cache: true
