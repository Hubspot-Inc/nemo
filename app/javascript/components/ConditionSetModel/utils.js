import ConditionSetModel from './ConditionSetModel';

/** Cache. */
const conditionSetStores = {};

/**
 * Returns a new instance of ConditionSetModel.
 *
 * Generally this should be added to a top-level Provider and used
 * once per condition set.
 */
export function provideConditionSetStore(uniqueId) {
  if (!conditionSetStores[uniqueId]) {
    conditionSetStores[uniqueId] = new ConditionSetModel();

    if (process.env.NODE_ENV === 'development') {
      // Debug helper.
      window.store = window.store || {};
      window.store[uniqueId] = conditionSetStores[uniqueId];
    }
  }

  return conditionSetStores[uniqueId];
}
