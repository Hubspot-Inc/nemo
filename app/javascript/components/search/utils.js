import isEmpty from 'lodash/isEmpty';
import queryString from 'query-string';

import FiltersModel from './FiltersModel';
import QuestionFilterModel from './QuestionFilterModel';

const MAX_HINTS_BEFORE_ELLIPSIZE = 1;

/**
 * Stringified controller_name from Rails.
 */
export const CONTROLLER_NAME = {
  RESPONSES: '"responses"',
};

/**
 * Returns a new instance of FiltersModel.
 *
 * Generally this should be added to a top-level Provider and only created once.
 */
export function createFiltersStore() {
  const store = new FiltersModel();

  if (process.env.NODE_ENV === 'development') {
    // Debug helper.
    if (window.store) {
      console.warn('WARN: Trying to create store that already exists.');
    }
    window.store = store;
  }

  return store;
}

/**
 * Returns a new instance of QuestionFilterModel.
 *
 * Generally this should be added to a top-level Provider and created
 * once per group of questions.
 */
export function createQuestionFilterStore(debugName) {
  const store = new QuestionFilterModel();

  if (process.env.NODE_ENV === 'development') {
    // Debug helper.
    window.store = window.store || {};
    if (window.store[debugName]) {
      console.warn('WARN: Trying to create store that already exists:', debugName);
    }
    window.store[debugName] = store;
  }

  return store;
}

/**
 * Given a list of hints (e.g. currently selected form names for the form filter button),
 * stringify them to be displayed on the button itself.
 */
export function getButtonHintString(hints) {
  if (isEmpty(hints)) {
    return '';
  }

  const joinedHints = hints.length > MAX_HINTS_BEFORE_ELLIPSIZE
    ? hints.length
    : hints.join(', ');

  return ` (${joinedHints})`;
}

/**
 * Given a form ID, find it in the list of all forms and return the name.
 */
export function getFormNameFromId(allForms, searchId) {
  const form = allForms.find(({ id }) => searchId === id);
  return (form && form.name) || 'Unknown';
}

/**
 * Given all of the different filter states,
 * return a stringified version for the backend.
 */
export function getFilterString({ allForms, selectedFormIds, advancedSearchText }) {
  const selectedFormNames = selectedFormIds
    .map((id) => JSON.stringify(getFormNameFromId(allForms, id)));

  const parts = [
    isEmpty(selectedFormNames) ? null : `form:(${selectedFormNames.join('|')})`,
    advancedSearchText,
  ].filter(Boolean);

  return parts.join(' ');
}

/**
 * Reload the page with the given search.
 */
export function submitSearch(filterString) {
  const parsed = queryString.parse(window.location.search);
  // The `search` query param will be removed from the URL if it's `undefined`.
  const search = filterString || undefined;
  const params = queryString.stringify({ ...parsed, search });

  window.location.assign(params
    ? `?${params}`
    : window.location.pathname);
}

/**
 * Returns true if the given param name exists and is non-empty.
 */
export function isQueryParamTruthy(paramName) {
  const parsed = queryString.parse(window.location.search);
  return Boolean(parsed[paramName]);
}
