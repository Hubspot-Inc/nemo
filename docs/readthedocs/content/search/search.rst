10. Search
~~~~~~~~~~

Searching is a critical aspect of being able to find the information you
need. This function is available on many parts of ELMO.

10.1. Operators
^^^^^^^^^^^^^^^

Keywords are just the words you use in your search. Combining Keywords
with Operators give the parameters for a search to occur. Keywords and
Operators form Expressions.

Operators in ELMO are: **AND, OR, NOT**\ ( ! or -), **grouping**
operator (parentheses), and **phrase** operator (“”).

+---+------------------------------------------------------------------+
| o | description                                                      |
| p |                                                                  |
| e |                                                                  |
| r |                                                                  |
| a |                                                                  |
| t |                                                                  |
| o |                                                                  |
| r |                                                                  |
+===+==================================================================+
| A | default implicit operator; matches when both of its arguments    |
| N | match example (with three keywords and two implicit AND          |
| D | operators between them): voters ballots stations returns matches |
|   | with voters AND ballots AND stations.                            |
+---+------------------------------------------------------------------+
| O | )                                                                |
| R |                                                                  |
| ( |                                                                  |
+---+------------------------------------------------------------------+
| N | Matches when the first argument matches, but the second one does |
| O | not. example: form != Closing returns the matches of forms that  |
| T | are NOT the Closing form example: ballot -box matches any        |
| ( | response with an answer containing the word ballot but NOT the   |
| ! | word box.                                                        |
| = |                                                                  |
| o |                                                                  |
| r |                                                                  |
| - |                                                                  |
| ) |                                                                  |
+---+------------------------------------------------------------------+
| ( | Grouping parenthesis explicitly denotes the argument boundaries. |
| … | example: (red                                                    |
| ) |                                                                  |
+---+------------------------------------------------------------------+
| “ | Quotes match when argument Keywords match an exact phrase.       |
| … | example: “The red fox jumped over the fence“ example: “Voter     |
| ” | lines went outside the center and down the street“               |
+---+------------------------------------------------------------------+


10.2. Qualifiers
^^^^^^^^^^^^^^^^

A qualifier is a word you add to an expression to specify where to
search. For example, searching **form: observation** within the
Responses menu will return all forms with the word “observation” in
them. Another example: searching **type: long** text in the Questions
menu returns all questions of the long text type.

Available qualifiers depend on the view or menu that you are working
within. They are listed below:

**Responses menu**

+----------+-----------------------------------------------------------+
| Qualifie | Function                                                  |
| r        |                                                           |
+==========+===========================================================+
| form:    | The name of the form submitted                            |
+----------+-----------------------------------------------------------+
| submitte | The name of the user that submitted the response (partial |
| r:       | matches allowed                                           |
+----------+-----------------------------------------------------------+
| submit-d | The date the response was submitted (e.g. 1985-03-22)     |
| ate:     |                                                           |
+----------+-----------------------------------------------------------+
| reviewed | Whether the response has been marked ‘reviewed’ (1 = yes  |
| :        | or 0 = no)                                                |
+----------+-----------------------------------------------------------+
| source:  | The medium via which the response was submitted (‘web’,   |
|          | ‘odk’, or ‘sms’)                                          |
+----------+-----------------------------------------------------------+
| text:    | Answers to textual questions                              |
+----------+-----------------------------------------------------------+

**Questions menu**

+-----+----------------------------------------------------------------+
| Qua | Function                                                       |
| lif |                                                                |
| ier |                                                                |
+=====+================================================================+
| cod | The question code (partial matches allowed)                    |
| e:  |                                                                |
+-----+----------------------------------------------------------------+
| tit | The question title (partial matches allowed)                   |
| le: |                                                                |
+-----+----------------------------------------------------------------+
| typ | The question type (text, long-text, integer, decimal,          |
| e:  | location, select-one, select-multiple, datetime, date, time)   |
+-----+----------------------------------------------------------------+
| tag | Tags applied to the question                                   |
| :   |                                                                |
+-----+----------------------------------------------------------------+

**Users menu**

+-----------------------------------+-----------------------------------+
| Qualifier                         | Function                          |
+===================================+===================================+
| name:                             | The user’s full name              |
+-----------------------------------+-----------------------------------+
| login:                            | The user’s username               |
+-----------------------------------+-----------------------------------+
| email:                            | The user’s email address          |
+-----------------------------------+-----------------------------------+
| phone:                            | The user’s phone number (no       |
|                                   | dashes or other punctuation,      |
|                                   | e.g. 1112223333                   |
+-----------------------------------+-----------------------------------+
| group:                            | The user group that the user      |
|                                   | belongs to                        |
+-----------------------------------+-----------------------------------+

**SMS Menu**

+-----+----------------------------------------------------------------+
| Qua | Function                                                       |
| lif |                                                                |
| ier |                                                                |
+=====+================================================================+
| con | The message content (partial matches allowed)                  |
| ten |                                                                |
| t:  |                                                                |
+-----+----------------------------------------------------------------+
| typ | The message type (incoming, reply, or broadcast) (partial      |
| e:  | matches allowed                                                |
+-----+----------------------------------------------------------------+
| use | The username of the sender or receiver (partial matches        |
| rna | allowed)                                                       |
| me: |                                                                |
+-----+----------------------------------------------------------------+
| nam | The full name of the sender or receiver (partial matches       |
| e:  | allowed)                                                       |
+-----+----------------------------------------------------------------+
| num | The phone number of the sender or receiver (partial matches    |
| ber | allowed)                                                       |
| :   |                                                                |
+-----+----------------------------------------------------------------+
| dat | The date the message was sent or received (e.g. 2015-01-29)    |
| e:  |                                                                |
+-----+----------------------------------------------------------------+
| dat | The date and time the message was sent or received (use        |
| eti | quotation marks and 24-hr time, e.g. “2015-01-29 14:00”)       |
| me: |                                                                |
+-----+----------------------------------------------------------------+