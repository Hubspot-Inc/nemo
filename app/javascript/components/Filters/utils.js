import isEmpty from 'lodash/isEmpty';
import mapKeys from 'lodash/mapKeys';
import queryString from 'query-string';
import moment from 'moment';
import { last } from 'lodash';

import { SUBMITTER_TYPES } from './SubmitterFilter/utils';

const MAX_HINTS_BEFORE_ELLIPSIZE = 1;

/**
 * Stringified controller_name from Rails.
 */
export const CONTROLLER_NAME = {
  RESPONSES: '"responses"',
};

/** Cache. */
let filtersStore = null;

/**
 * Returns a new instance of FiltersModel.
 *
 * Generally this should be added to a top-level Provider and only used once.
 */
export function provideFiltersStore(FiltersModel, initialState) {
  if (!filtersStore) {
    filtersStore = new FiltersModel(initialState);

    if (process.env.NODE_ENV === 'development') {
      // Debug helper.
      window.store = filtersStore;
    }
  }

  return filtersStore;
}

/**
 * Given a list of hints (e.g. currently selected form names for the form filter button),
 * stringify them to be displayed on the button itself.
 */
export function getButtonHintString(hints) {
  if (isEmpty(hints)) {
    return '';
  }

  const joinedHints = hints.length > MAX_HINTS_BEFORE_ELLIPSIZE
    ? `${hints.length} filters`
    : hints.join(', ');

  return ` (${joinedHints})`;
}

/**
 * Given an item ID, find it in the list of items and return its `name`.
 */
export function getItemNameFromId(allItems, searchId, nameKey = 'name') {
  const item = allItems.find(({ id }) => searchId === id);
  return (item && item[nameKey]) || 'Unknown';
}

/**
 * Given a leftQingId, find it in the list of all questions and return the name.
 */
export function getQuestionNameFromId(allQuestions, searchId) {
  return getItemNameFromId(allQuestions, searchId, 'code');
}

/**
 * Converts a list of data from the backend into something Select2 understands, e.g.
 * [{ id: '1', name: 'One' }, ...] => [{ id: '1', text: 'One' }, ...]
 */
export function parseListForSelect2(allItems) {
  return allItems.map((item) =>
    mapKeys(item, (value, key) => (key === 'name' ? 'text' : key)));
}

/**
 * Given all of the different filter states,
 * return a stringified version for the backend.
 */
export function getFilterString({
  selectedFormIds,
  conditionSetStore,
  isReviewed,
  selectedSubmittersForType,
  advancedSearchText,
  startDate,
  endDate,
}) {
  const allQuestions = conditionSetStore.refableQings;
  const questionFilters = conditionSetStore.conditions
    .filter(({ leftQingId, currTextValue, remove }) => leftQingId && currTextValue && !remove)
    .map(({ leftQingId, currTextValue }) =>
      `{${getQuestionNameFromId(allQuestions, leftQingId)}}:${JSON.stringify(currTextValue)}`);

  const submitterParts = SUBMITTER_TYPES.map((type) => {
    const selectedSubmitterIds = selectedSubmittersForType[type].map(({ id }) => id);
    return isEmpty(selectedSubmitterIds) ? null : `${type}-id:(${selectedSubmitterIds.join('|')})`;
  });

  // Only include one upper bound and one lower bound date. Use widest range.
  let searchDates = advancedSearchText.split(' ').filter((s) => s.indexOf('submit-date') !== -1).map(queryToMoment);
  const searchText = advancedSearchText.split(' ').filter((s) => s.indexOf('submit-date') === -1).join(' ');
  searchDates = [...searchDates, startDate, endDate].filter((d) => d !== null);
  if (searchDates.length >= 2) {
    searchDates.sort((a, b) => {
      if (a.isSameOrBefore(b)) return -1;
      return 1;
    });
  }

  let dateQueries = null;
  if (searchDates.length === 0) {
    // No-op
  } else if (searchDates.length === 1) {
    // Figure out if this was supposed to be an upper or lower bound.
    if (searchDates[0] === startDate) {
      dateQueries = `submit-date >= ${searchDates[0].format('YYYY-MM-DD')}`;
    } else if (searchDates[0] === endDate) {
      dateQueries = `submit-date <= ${searchDates[0].format('YYYY-MM-DD')}`;
    } else {
      // Date entered in search bar, use original string.
      dateQueries = advancedSearchText.split(' ').filter((s) => s.indexOf('submit-date') !== -1).join(' ');
    }
  } else {
    // Take first and last dates.
    dateQueries = `submit-date >= ${searchDates[0].format('YYYY-MM-DD')} submit-date <= ${last(searchDates).format('YYYY-MM-DD')}`;
  }

  const parts = [
    isEmpty(selectedFormIds) ? null : `form-id:(${selectedFormIds.join('|')})`,
    ...questionFilters,
    isReviewed == null ? null : `reviewed:${isReviewed ? '1' : '0'}`,
    ...submitterParts,
    searchText,
    dateQueries,
  ].filter(Boolean);

  return parts.join(' ');
}

/**
 * Reload the page with the given search.
 */
export function submitSearch(filterString) {
  const parsed = queryString.parse(window.location.search);
  // The `search` query param will be removed from the URL if it's `undefined`.
  const search = filterString || undefined;
  const params = queryString.stringify({ ...parsed, search });

  window.location.assign(params
    ? `?${params}`
    : window.location.pathname);
}

/**
 * Returns true if the given param name exists and is non-empty.
 */
export function isQueryParamTruthy(paramName) {
  const parsed = queryString.parse(window.location.search);
  return Boolean(parsed[paramName]);
}

export function queryToMoment(query) {
  if (query === null || query === undefined) return null;
  const dateString = query.split(/submit-date[<>=]+/)[1];
  // If this doesn't parse an actual date, don't use it.
  if (dateString === undefined) return null;
  const date = moment(dateString);
  if (date._isValid) { /* eslint-disable-line no-underscore-dangle */
    return date;
  }
  return null;
}
