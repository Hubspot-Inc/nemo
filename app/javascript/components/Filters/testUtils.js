import FiltersModel from './model';
import ConditionSetModel from '../ConditionSetFormField/model';
import { submitterType } from './SubmitterFilter/component';

export const getFiltersStore = () => new FiltersModel({
  conditionSetStore: new ConditionSetModel({
    conditions: [
      {
        leftQingId: '1',
        op: 'eq',
        operatorOptions: [
          { name: '= equals', id: 'eq' },
          { name: 'â‰  does not equal', id: 'neq' },
        ],
        value: '5',
      },
    ],

    refableQings: [
      {
        id: '1',
        code: 'One',
        fullDottedRank: '1',
        rank: 1,
      },
      {
        id: '2',
        code: 'Two',
        fullDottedRank: '2',
        rank: 2,
      },
    ],
  }),

  allForms: [
    { id: '1', name: 'One' },
    { id: '2', name: 'Two' },
    { id: '3', name: "Q\"uo'te`s" },
  ],

  originalFormIds: [
    '1',
  ],

  selectedFormIds: [
    '2',
  ],

  originalIsReviewed: null,

  isReviewed: true,

  allSubmittersForType: {
    [submitterType.USER]: [
      { id: 'A', name: 'User A' },
      { id: 'B', name: 'User B' },
    ],
    [submitterType.GROUP]: [
      { id: 'A', name: 'Group A' },
      { id: 'B', name: 'Group B' },
    ],
  },

  originalSubmittersForType: {
    [submitterType.USER]: [{ id: 'A', name: 'User A' }],
    [submitterType.GROUP]: [],
  },

  selectedSubmittersForType: {
    [submitterType.USER]: [{ id: 'B', name: 'User B' }],
    [submitterType.GROUP]: [{ id: 'B', name: 'Group B' }],
  },

  advancedSearchText: 'query',
});
