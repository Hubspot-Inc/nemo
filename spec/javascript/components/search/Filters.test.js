import React from 'react';
import { shallow, mount } from 'enzyme';

import { STUB_COMPONENT_WARNINGS, suppressErrors, unsuppressAllErrors } from '../../testUtils';
import { allFilterProps } from './utils';

import { CONTROLLER_NAME } from '../../../../app/javascript/components/search/utils';
import Component from '../../../../app/javascript/components/search/Filters';

const defaultProps = {
  ...allFilterProps,
  controllerName: CONTROLLER_NAME.RESPONSES,
};

it('renders as expected (responses page)', () => {
  const wrapper = shallow(<Component {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

it('renders as expected (other page)', () => {
  const wrapper = shallow(
    <Component
      {...defaultProps}
      controllerName="foo"
    />,
  );
  expect(wrapper).toMatchSnapshot();
});

describe('integration', () => {
  suppressErrors(STUB_COMPONENT_WARNINGS);
  const wrapper = mount(<Component {...defaultProps} />);
  unsuppressAllErrors();

  beforeEach(() => window.location.assign.mockClear());

  it('navigates on apply form filter', () => {
    wrapper.find('Button.btn-form-filter').simulate('click');
    const overlay = shallow(wrapper.find('OverlayTrigger').prop('overlay'));
    overlay.find('Select2').simulate('change', { target: { value: defaultProps.allForms[0].id } });

    expect(window.location.assign).toMatchSnapshot();
    overlay.find('Button.btn-apply').simulate('click');
    expect(window.location.assign).toMatchSnapshot();
  });

  it('navigates on apply advanced search', () => {
    wrapper.find('.search-str').simulate('change', { target: { value: 'something else' } });

    expect(window.location.assign).toMatchSnapshot();
    wrapper.find('Button.btn-advanced-search').simulate('click');
    expect(window.location.assign).toMatchSnapshot();
  });
});
