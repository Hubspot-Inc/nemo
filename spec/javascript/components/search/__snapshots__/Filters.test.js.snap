// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration navigates on apply advanced search 1`] = `[MockFunction]`;

exports[`integration navigates on apply advanced search 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "?search=form%3A%28%22Two%22%29%20%7BOne%7D%3A%225%22%20reviewed%3A1%20something%20else",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`integration navigates on apply form filter 1`] = `[MockFunction]`;

exports[`integration navigates on apply form filter 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "?search=form%3A%28%22Two%22%29%20%7BOne%7D%3A%225%22%20reviewed%3A1%20query",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`integration navigates on apply question filter 1`] = `[MockFunction]`;

exports[`integration navigates on apply question filter 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "?search=form%3A%28%22Two%22%29%20%7BOne%7D%3A%225%22%20reviewed%3A1%20query",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`integration navigates on apply reviewed filter 1`] = `[MockFunction]`;

exports[`integration navigates on apply reviewed filter 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "?search=form%3A%28%22Two%22%29%20%7BOne%7D%3A%225%22%20reviewed%3A1%20query",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`renders as expected (other page) 1`] = `
<div
  className="filters"
>
  <AdvancedSearchFilter
    onClear={[Function]}
    onSubmit={[Function]}
  />
</div>
`;

exports[`renders as expected (responses page) 1`] = `
<div
  className="filters"
>
  <ButtonToolbar>
    <FormFilter
      onSubmit={[Function]}
    />
    <QuestionFilter
      onSubmit={[Function]}
    />
    <FormFilter
      onSubmit={[Function]}
    />
  </ButtonToolbar>
  <AdvancedSearchFilter
    onClear={[Function]}
    onSubmit={[Function]}
  />
</div>
`;
